src/Arch.SourceGen/VariadicGenerator.cs

```diff
@@ -0,0 +1,252 @@
+﻿using System.Text.RegularExpressions;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+namespace Arch.SourceGen;
+
+/// <summary>
+///     The <see cref="VariadicGenerator"/> finds assembly members with <see cref="Arch.Core.VariadicAttribute"/>, and generates variadic types.
+///     See the attribute's documentation for information on usage.
+/// </summary>
+[Generator]
+public class VariadicGenerator : IIncrementalGenerator
+{
+    // A note on implementation:
+    // This class is implemented on a line-by-line basis making heavy use of regular expressions to parse C# code.
+    // Clearly, this isn't ideal. The "correct" way of implementing something like this would be a CSharpSyntaxRewriter.
+    // However, CSharpSyntaxRewriter is poorly documented and can lead to a lot of very very difficult code.
+    // For now, since this is limited to Arch, it's enough to use this.
+
+    private class VariadicInfo
+    {
+        public string Type { get; set; } = string.Empty;
+        public string Code { get; set; } = string.Empty;
+        public string Name { get; set; } = string.Empty;
+        public string EnclosingType { get; set; } = string.Empty;
+        public string EnclosingTypeName { get; set; } = string.Empty;
+        public string Namespace { get; set; } = string.Empty;
+        public string Usings { get; set; } = string.Empty;
+        public int Count { get; set; } = 0;
+        public int Start { get; internal set; }
+    }
+
+    public void Initialize(IncrementalGeneratorInitializationContext context)
+    {
+        // Uncomment to help debug with breakpoints:
+        // Debugger.Launch();
+
+        var infos = context.SyntaxProvider.ForAttributeWithMetadataName("Arch.Core.VariadicAttribute",
+            // We want to grab everything with the attribute, always!
+            (node, token) => true,
+            (ctx, token) =>
+            {
+                VariadicInfo info = new();
+
+                // Note: doesn't support mulilevel type nesting, like struct Something { struct AnotherThing { struct AThirdThing {} } }. Only 1 nesting layer is supported.
+                if (ctx.TargetSymbol.ContainingType is not null)
+                {
+                    info.EnclosingTypeName = ctx.TargetSymbol.ContainingType.Name;
+                    var accessibility = ctx.TargetSymbol.ContainingType.DeclaredAccessibility switch
+                    {
+                        Accessibility.Public => "public",
+                        _ => "internal",
+                    };
+
+                    var reference = ctx.TargetSymbol.ContainingType.IsReferenceType ? "class" : "struct";
```

Comment:
Records are also reference types. This works for now but I believe a future version of C# will introduce record structs.

```suggestion
                    var reference = ctx.TargetSymbol.ContainingType.IsRecord ? "record" :  ctx.TargetSymbol.ContainingType.IsValueType ? "struct" : "class";
```

Reply:
Record structs actually exist in C#10 already. I'll make sure those are supported.

Reply:
You're right! I am behind.